File: lib/rest/yampi/mappers/yampi_product_mapper.dart
import 'package:sertton/core/catalog/dtos/product_dto.dart';
import 'package:sertton/core/global/responses/pagination_response.dart';
import 'package:sertton/rest/yampi/mappers/yampi_brand_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_sku_mapper.dart';
import 'package:sertton/rest/types/json.dart';

class YampiProductMapper {
  static ProductDto toDto(Json json) {
    if (json.containsKey('data')) {
      json = json['data'];
    }
    String imageUrl = '';
    final images = json['images'];
    if (images != null && images['data'] != null) {
      final imagesList = images['data'] as List;
      if (imagesList.isNotEmpty) {
        final firstImage = imagesList[0] as Map<String, dynamic>;
        imageUrl =
            firstImage['large']?['url'] ??
            firstImage['medium']?['url'] ??
            firstImage['thumb']?['url'] ??
            firstImage['small']?['url'] ??
            '';
      }
    }

    String description = '';
    String specifications = '';
    final texts = json['texts'];
    if (texts != null && texts['data'] != null) {
      description = texts['data']['description'] ?? '';
      specifications = texts['data']['specifications'] ?? '';
    }

    return ProductDto(
      id: json['id'].toString(),
      name: json['name'] ?? '',
      slug: json['slug'] ?? '',
      description: description,
      specifications: specifications,
      imageUrl: imageUrl,
      skuCode: '${json['sku']}'.split(',').first,
      brand: YampiBrandMapper.toDto(json['brand']['data']),
      skus: YampiSkuMapper.toDtoList(json['skus']),
    );
  }

  static List<ProductDto> toDtoList(Json json) {
    final data = (json['data'] as List).cast<Json>();
    return data.map(toDto).toList();
  }

  static PaginationResponse<ProductDto> toDtoPagination(Json json) {
    final meta = json['meta']['pagination'] as Json;
    return PaginationResponse(
      items: toDtoList(json),
      itemsPerPage: meta['per_page'],
      currentPage: meta['current_page'],
      totalItems: meta['total'],
    );
  }
}

File: sertton-react-native-app/src/rest/yampi/mappers/yampi-product-mapper.ts
import type { Product } from "@/core/catalog/entities"
import { createPagination, type Pagination } from "@/core/shared/responses/pagination"
import type { YampiProduct, YampiResponse } from "@/rest/yampi/types"
import { YampiBrandMapper } from "./yampi-brand-mapper"
import { YampiSkuMapper } from "./yampi-sku-mapper"

export const YampiProductMapper = () => {
  const brandMapper = YampiBrandMapper()
  const skuMapper = YampiSkuMapper()

  const toDomain = (input: YampiProduct): Product => {
    const image = input.images.data[0]
    const imageUrl =
      image?.large?.url ?? image?.medium?.url ?? image?.thumb?.url ?? image?.small?.url ?? ""

    return {
      id: String(input.id),
      name: input.name,
      slug: input.slug,
      description: input.texts.data.description,
      specifications: input.texts.data.specifications,
      imageUrl,
      skuCode: input.sku.split(",")[0] ?? "",
      brand: brandMapper.toDomain(input.brand.data),
      skus: input.skus.data.map(skuMapper.toDomain),
    }
  }

  return {
    toDomain,
    toPagination(input: YampiResponse<YampiProduct>): Pagination<Product> {
      const pagination = input.meta?.pagination
      return createPagination({
        currentPage: pagination?.current_page ?? 1,
        items: input.data
          .map(toDomain)
          .filter((product) => product.skus.some((sku) => sku.salePrice > 0)),
        itemsPerPage: pagination?.per_page ?? input.data.length,
        totalItems: pagination?.total ?? input.data.length,
      })
    },
  }
}
